hail/src/main/scala/is/hail/expr/ir/Binds.scala

```diff
@@ -1,36 +1,89 @@
+    case TableFilter(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case TableMapGlobals(child, _) => if (i == 1) child.typ.globalEnv.m else empty
+    case TableMapRows(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case TableAggregateByKey(child, _) => if (i == 1) child.typ.globalEnv.m else empty
+    case TableKeyByAndAggregate(child, _, _, _, _) => if (i == 1) child.typ.globalEnv.m else if (i == 2) child.typ.rowEnv.m else empty
+    case MatrixMapRows(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case MatrixFilterRows(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case MatrixMapCols(child, _, _) => if (i == 1) child.typ.colEnv.m else empty
+    case MatrixFilterCols(child, _) => if (i == 1) child.typ.colEnv.m else empty
+    case MatrixMapEntries(child, _) => if (i == 1) child.typ.entryEnv.m else empty
+    case MatrixFilterEntries(child, _) => if (i == 1) child.typ.entryEnv.m else empty
+    case MatrixMapGlobals(child, _) => if (i == 1) child.typ.globalEnv.m else empty
+    case MatrixAggregateColsByKey(child, _, _) => if (i == 1) child.typ.rowEnv.m else if (i == 2) child.typ.globalEnv.m else empty
+    case MatrixAggregateRowsByKey(child, _, _) => if (i == 1) child.typ.colEnv.m else if (i == 2) child.typ.globalEnv.m else empty
+    case _ => empty
+  }
+}
+
+object AggBindings {
+  private val empty: Array[(String, Type)] = Array()
+
+  def apply(x: BaseIR, i: Int): Iterable[(String, Type)] = x match {
+    case AggLet(name, value, _) => if (i == 1) Array(name -> value.typ) else empty
+    case AggExplode(a, name, _) => if (i == 1) Array(name -> a.typ.asInstanceOf[TContainer].elementType) else empty
+    case AggArrayPerElement(a, name, _) => if (i == 1) Array(name -> a.typ.asInstanceOf[TContainer].elementType) else empty
+    case ArrayAgg(a, name, _) => if (i == 1) Array(name -> a.typ.asInstanceOf[TContainer].elementType) else empty
+    case TableAggregate(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case MatrixAggregate(child, _) => if (i == 1) child.typ.entryEnv.m else empty
+    case TableAggregateByKey(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case TableKeyByAndAggregate(child, _, _, _, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case MatrixMapRows(child, _) => if (i == 1) child.typ.entryEnv.m else empty
+    case MatrixMapCols(child, _, _) => if (i == 1) child.typ.entryEnv.m else empty
+    case MatrixAggregateColsByKey(child, _, _) => if (i == 1) child.typ.entryEnv.m else if (i == 2) child.typ.colEnv.m else empty
+    case MatrixAggregateRowsByKey(child, _, _) => if (i == 1) child.typ.entryEnv.m else if (i == 2) child.typ.rowEnv.m else empty
+    case _ => empty
   }
 }
+
+object ChildEnv {
+  private val empty = (Env.empty[Type], None)
+
+  def apply(ir: BaseIR, i: Int, env: Env[Type], aggEnv: Option[Env[Type]]): (Env[Type], Option[Env[Type]]) = ir match {
```

Comment:
I would still like to see this composed of two pieces.  I think the piece that propagates the bindings downward can be defined in terms of `UsesAggEnv`.

Reply:
I don't understand what you mean.

Reply:
Consider an IR: (X c), where X is some operation and c is some child.  X is evaluated in some environment env/aggEnv (if it is a value IR).  I was suggesting a function that computes the part of that environment visible to a (without bindings).  Say, ChildEnvWithoutBindings.  Then ChildEnv is just ChildEnvWithoutBindings plus the bindings.

This is all a bit gnarly so I'm trying to identify simple, understandable pieces we can rely on.

Reply:
Ah, got it. Sure.

Reply:
OK, is this what you had in mind?